analysers/analyser_osmosis_highway_deadend.py

```diff
@@ -206,6 +206,48 @@
   r.nid IS NULL
 """
 
+
+sql40 = """
+SELECT
+    wid,
+    nid,
+    ST_AsText(geom)
+FROM (
+SELECT
+    MIN(way_ends.id) AS wid,
+    nodes.id AS nid,
+    nodes.geom,
+    MIN(way_ends.highway) AS highway
+FROM
+    {0}highway_ends AS way_ends
+    JOIN highways ON
+        highways.linestring && way_ends.linestring AND
+        way_ends.nid = ANY(highways.nodes)
+    JOIN nodes ON
+        nodes.id = way_ends.nid AND
+        (NOT nodes.tags?'highway' OR (
+          nodes.tags->'highway' != 'turning_circle' AND
+          nodes.tags->'highway' != 'turning_loop' AND
+          nodes.tags->'highway' != 'mini_roundabout'
+        )) AND
+        (NOT nodes.tags?'entrance' OR nodes.tags->'entrance' = 'no') -- i.e. indoor part not drawn
+WHERE
+    way_ends.highway = 'service'
+GROUP BY
+    nodes.id,
+    nodes.geom
+HAVING
+    COUNT(*) = 1
+) AS t
+    LEFT JOIN highways AS drivethrough ON
```

Comment:
Need to exclude the highway it self, need not it?

```sql
drivethrough.id = t.wid
```

Reply:
I've added the check, thank you!

Do I understand correctly that this would be needed in case `t` is a connection between a service road and a non-service road (like residential)?

Reply:
My impression was that `t` already contains only nodes that are not connected to any other highways (and not turning circles and such)

Here I indeed just wanted to have a filter on drive-throughs. As far as I understand the code that generates `highway_ends` (and thus `t`), `t` doesn't contain the tags, correct?

But as mentioned, I'm really a beginner in SQL

Reply:
(Note that I do have some concerns about the WHERE block in the inner FROM, i.e. are connections between service and unclassified filtered out of `t` or not)

Reply:
So. There is a more efficient way to do this.
The idea is to start with `highways` as highways_services filtered on required services, join on `highway_ends`, then join on all (other) highways and then join on `nodes`. It can filter the checked highways from start of the process.

On a such complex query you can ignore `{0}` and the `.format` part.

To work on this, run osmose once with --no-clean option and jump to the database to get postgres shell to test query directly. Once here you can run query, with timing (`\timing`) and more important `ANALYSER` (`ANALYSER SELECT ...`).

Reply:
I'm afraid I'll have to park this for a while then, until I figure out how to run the SQL queries locally.

Reply:
Look at this https://github.com/osm-fr/osmose-backend/tree/master/docker#develop-on-osmose-using-docker

Reply:
Oh wow, it took me 3 hours to discover that in 
`./osmose_run.py --no-clean --country=monaco --analyser=osmosis_highway_deadend`
you have to use the filename *excluding* the prefix `analyser_`...
And I haven't noticed that any earlier time that I tried to build it either...

Reply:
Thank you very much for the feedback earlier!
Now that I can run them locally, I rewrote the tests; tested them, and they seem to work great on the 7 areas where I ran them.
I'll commit them in a new pull request to allow more easy comparison of what has changed from the original code: #1528